node/storage.go

```diff
@@ -0,0 +1,184 @@
+package node
+
+import (
+	"fmt"
+
+	"github.com/filecoin-project/storetheindex/store"
+	"github.com/ipfs/go-cid"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+)
+
+var _ store.Storage = &nodeStorage{}
+
+// nodeStorage combines a cache and a persistent storage
+type nodeStorage struct {
+	primary    store.Storage
+	persistent store.PersistentStorage
+}
+
+// NewStorage creates new nodeStorage from a primary and persistent storages
+func NewStorage(primary store.Storage, persistent store.PersistentStorage) *nodeStorage {
+	return &nodeStorage{primary, persistent}
+}
+
+// Get retrieves IndexEntries for a CID
+func (ns *nodeStorage) Get(c cid.Cid) ([]store.IndexEntry, bool, error) {
+	if ns.primary != nil {
+		// Check if CID in primary storage
+		v, found, err := ns.primary.Get(c)
+
+		if ns.persistent != nil && (!found || err != nil) {
+			v, found, err = ns.persistent.Get(c)
+			// TODO: What about adding a CacheStorage interface that includes
+			// putEntries(cid, []IndexEntry) function
+			// so we don't need to loop through IndexEntry to move from
+			// one storage to another?
+			if found && err == nil {
+				// Move from persistent to cache
+				for i := range v {
+					ns.primary.Put(c, v[i])
+				}
+			}
+		}
+		return v, found, err
+	}
+
+	// If no primary storage, get from persistent
+	return ns.persistent.Get(c)
+}
+
+// Put stores entry for a CID if the entry is not already
+// stored.  New entries are added to the entries that are already there.
+func (ns *nodeStorage) Put(c cid.Cid, entry store.IndexEntry) (bool, error) {
+	return ns.put(c, entry)
+}
+
+func (ns *nodeStorage) put(c cid.Cid, entry store.IndexEntry) (bool, error) {
+	var ok bool
+	var err error
+	// If there is persistent storage we put in persistent storage
```

Comment:
Should the primary be checked first?  If there is a guarantee that anything that is in the primary is already in persistent storage, then that would avoid unneeded updates without having to query persistent storage, for values already cached.

Doing this would probably require a new API, `HasValue(cid.Cid, store.IndexEntry) bool` or would mean looking at the value(s) returned by `Get`.  Is it worth it?

Reply:
You are right. This is a better approach. Fixed. Thanks!